src/main/java/org/testng/internal/Parameters.java

```diff
@@ -264,35 +265,81 @@ else if (ITestResult.class.equals(c)) {
-        String p = parameterNames[j];
-        String value = params.xmlParameters.get(p);
-        if (null == value) {
-          // try SysEnv entries
-          value = System.getProperty(p);
+    if (isInjected) {
+      FilterOutInJectedTypesResult filterOutResult = FilterOutInJectedTypesFromOptionalValues(parameterTypes, optionalValues);
+      optionalValues = filterOutResult.getOptionalValues();
+      parameterTypes = filterOutResult.getParameterTypes();
+    }
+    for (int i = 0; i < parameterNames.length; i++) {
+      String p = parameterNames[i];
+      String value = params.xmlParameters.get(p);
+      if (null == value) {
+        // try SysEnv entries
+        value = System.getProperty(p);
+      }
+      if (null == value) {
+        if (optionalValues != null) {
+            value = optionalValues[i];
         }
         if (null == value) {
-          if (optionalValues != null) {
-            value = optionalValues[i];
-          }
-          if (null == value) {
             throw new TestNGException("Parameter '" + p + "' is required by "
                     + methodAnnotation
                     + " on " + prefix + " "
                     + name
                     + " but has not been marked @Optional or defined\n"
                     + (xmlSuite.getFileName() != null ? "in "
-                    + xmlSuite.getFileName() : ""));
-          }
+                            + xmlSuite.getFileName() : ""));
+        }
+    }
+      vResult.add(convertType(parameterTypes[i], value, p));
+    }
+    return vResult;
+  }
+  
+  /**
+   * Remove injected types from parameterTypes and optionalValues
+   * @param parameterTypes
+   * @param optionalValues
+   * @return FilterOutInJectedTypesResult
+   */
+  private static FilterOutInJectedTypesResult FilterOutInJectedTypesFromOptionalValues(Class<?>[] parameterTypes, String[] optionalValues) {
+    List<Class<?>> typeList = Lists.newArrayList();
+    List<String> optionalValueList = Lists.newArrayList();
+    Collections.addAll(typeList, parameterTypes);
+    Collections.addAll(optionalValueList, optionalValues);
+    Iterator<Class<?>> typeIterator = typeList.iterator();
+    Iterator<String> optionalIterator = optionalValueList.iterator();
+    while (typeIterator.hasNext()) {
+        Class<?> parameterType = typeIterator.next();
+        optionalIterator.next();
+        if (INJECTED_TYPES.contains(parameterType)) {
+            optionalIterator.remove();
+            typeIterator.remove();
         }
+    }
+    return new FilterOutInJectedTypesResult(typeList.toArray(new Class<?>[0]),
+            optionalValueList.toArray(new String[0]));
+  }
 
-        vResult.add(convertType(parameterTypes[i], value, p));
-        j++;
-      }
+  /**
+   * Store the result of parameterTypes and optionalValues after filter out injected types
+   */
+  private final static class FilterOutInJectedTypesResult {
+    private Class<?>[] parameterTypes;
```

Comment:
Prefer `ArrayList` over arrays.

Reply:
You mean:
Change 
`optionalValues = filterOutResult.getOptionalValues();`
to 
`optionalValues = filterOutResult.getOptionalValues().toArray(new Class<?>[0]);`
in 
`Parameters::createParam`?